pkg/cmd/roachtest/tests/disk_stall.go

```diff
@@ -15,125 +15,441 @@ import (
+	ctx context.Context, t test.Test, c cluster.Cluster, nodeID int,
+) (time.Duration, bool) {
+	return getProcessMonotonicTimestamp(ctx, t, c, nodeID, "ActiveExitTimestampMonotonic")
+}
 
-	if expectMsg != strings.Contains(result.Stderr, "disk stall detected") {
-		t.Fatalf("unexpected output: %v", result.Err)
-	} else if elapsed := timeutil.Since(tStarted); !expectMsg && elapsed < dur {
-		t.Fatalf("no disk stall injected, but process terminated too early after %s (expected >= %s)", elapsed, dur)
+func getProcessMonotonicTimestamp(
+	ctx context.Context, t test.Test, c cluster.Cluster, nodeID int, prop string,
+) (time.Duration, bool) {
+	details, err := c.RunWithDetailsSingleNode(ctx, t.L(), c.Node(nodeID),
+		"systemctl show cockroach.service --property="+prop)
+	if err != nil {
+		t.Fatal(err)
+	} else if details.Err != nil {
+		t.Fatal(details.Err)
+	}
+	parts := strings.Split(details.Stdout, "=")
+	if len(parts) < 2 {
+		return 0, false
+	}
+	s := strings.TrimSpace(parts[1])
+	if s == "" {
+		return 0, false
+	}
+	u, err := strconv.ParseUint(s, 10, 64)
+	if err != nil {
+		t.Fatalf("unable to parse monotonic timestamp %q: %s", parts[1], err)
 	}
+	if u == 0 {
+		return 0, false
+	}
+	return time.Duration(u) * time.Microsecond, true
+}
+
+type diskStaller interface {
+	Setup(ctx context.Context)
+	Cleanup(ctx context.Context)
+	Stall(ctx context.Context, nodes option.NodeListOption)
+	Unstall(ctx context.Context, nodes option.NodeListOption)
+	DataDir() string
+	LogDir() string
+}
+
+type dmsetupDiskStaller struct {
+	t test.Test
+	c cluster.Cluster
+}
 
-	c.Run(ctx, n, "charybdefs-nemesis --clear")
-	c.Run(ctx, n, "sudo umount {store-dir}/faulty")
+var _ diskStaller = (*dmsetupDiskStaller)(nil)
+
+func (s *dmsetupDiskStaller) device() string { return getDevice(s.t, s.c.Spec()) }
+
+func (s *dmsetupDiskStaller) Setup(ctx context.Context) {
+	dev := s.device()
+	s.c.Run(ctx, s.c.All(), `sudo umount /mnt/data1`)
+	s.c.Run(ctx, s.c.All(), `sudo dmsetup remove_all`)
+	s.c.Run(ctx, s.c.All(), `echo "0 $(sudo blockdev --getsz `+dev+`) linear `+dev+` 0" | `+
+		`sudo dmsetup create data1`)
+	s.c.Run(ctx, s.c.All(), `sudo mount /dev/mapper/data1 /mnt/data1`)
+}
+
+func (s *dmsetupDiskStaller) Cleanup(ctx context.Context) {
+	s.c.Run(ctx, s.c.All(), `sudo umount /mnt/data1`)
+	s.c.Run(ctx, s.c.All(), `sudo dmsetup remove_all`)
+	s.c.Run(ctx, s.c.All(), `sudo mount /mnt/data1`)
+}
+
+func (s *dmsetupDiskStaller) Stall(ctx context.Context, nodes option.NodeListOption) {
+	s.c.Run(ctx, nodes, `sudo dmsetup suspend --noflush --nolockfs data1`)
+}
+
+func (s *dmsetupDiskStaller) Unstall(ctx context.Context, nodes option.NodeListOption) {
+	s.c.Run(ctx, nodes, `sudo dmsetup resume data1`)
+}
+
+func (s *dmsetupDiskStaller) DataDir() string { return "{store-dir}" }
+func (s *dmsetupDiskStaller) LogDir() string  { return "logs" }
+
+type cgroupDiskStaller struct {
+	t           test.Test
+	c           cluster.Cluster
+	readOrWrite []string
+	logsToo     bool
```

Comment:
We should have tests where we stall the log device independently of the data device.

Reply:
Agreed, but it's a little tricky given the volume-level stalling. The FUSE staller makes it a little easier and does implement stalling of the log device independently. @nicktrav will merge his change to fixup and unskip them after this goes in.